### ISSUE TITLE
How to get Griddle state outside of component?

### ISSUE THREAD

**Issue by MichaelDeBoey** [GriddleGriddle/Griddle#796]
## Griddle version
`1.11.1`

## Question
How can I get the Griddle state from outside the component?

Like you can see in [the CodeSandbox](https://codesandbox.io/s/j702609rn9?expanddevtools=1&module=%2FGridInfo.js), I want to use (for testing purposes, just show) the `filter` value **outside** the Griddle component.
Is there a recommended way of doing this?

CC: @dahlbyk

**Comment by dahlbyk** [comment]
@MichaelDeBoey you could inject an event handler into Griddle to receive filter updates, but there's not a great way to push an external filter update back into Griddle's store.

More common would be to replace Griddle's `Filter` with a component connected to an external store. Check out [this story](https://github.com/GriddleGriddle/Griddle/blob/250ea2a019895ee490e9f70e1ca26b4dfb9e181a/stories/index.tsx#L968-L1018).

**Comment by MichaelDeBoey** [comment]
@dahlbyk The goal is to really remove the `Filter` within the `Griddle` component and have a custom `Filter` on a totally different place outside the `Griddle` component

**Comment by MichaelDeBoey** [comment]
@dahlbyk The goal is to really remove the `Filter` within the `Griddle` component and have a custom `Filter` on a totally different place outside the `Griddle` component

Any idea how this could be accomplished?

**Comment by dahlbyk** [comment]
That same story basically does this already - `data` is filtered outside the `<Griddle />`. Just set `Filter: () => null` to replace the default `Filter`.

**Comment by MichaelDeBoey** [comment]
@dahlbyk That story also filters the data, I was thinking about just setting the filter value and let `Griddle` do their filtering, like they would do with the internal `Filter`

**Comment by dahlbyk** [comment]
Try this modification to that story:

```diff
diff --git a/stories/index.tsx b/stories/index.tsx
index 6fe219f..191b8ce 100644
--- a/stories/index.tsx
+++ b/stories/index.tsx
@@ -985,15 +985,18 @@ storiesOf('Griddle main', module)
       })
     )(CustomFilterComponent);
 
+
     const plugins = [
       LocalPlugin,
       {
-        components: { Filter: CustomFilterConnectedComponent },
+        components: { Filter: () => null },
       },
     ];
     const SomePage = props => (
       <div>
-        <Griddle data={props.data} plugins={plugins} storeKey="griddleStore" />
+        <Griddle data={props.data} filter={props.filter} plugins={plugins} storeKey="griddleStore" />
 
         Component outside of Griddle that's sharing state
         <CustomFilterConnectedComponent />
@@ -1002,9 +1005,8 @@ storiesOf('Griddle main', module)
 
     const SomePageConnected = reduxConnect(
       (state: TestState) => ({
-        data: !state.searchString ? state.data :
-          state.data.filter(r =>
-            Object.keys(r).some(k => r[k] && r[k].toString().indexOf(state.searchString) > -1)),
+        data: state.data,
+        filter: state.searchString,
       })
     )(SomePage);
```

Ultimately I don't know that you'll be able to use `actions.setFilter` to do what you want, as you won't be able to access Griddle's internal Redux store (at least not without breaking with React best practices). But if you track `filter` externally, providing it through `props` seems to do the trick.

**Comment by MichaelDeBoey** [comment]
@dahlbyk That seems to work indeed, but like you said that's a bit of a "hack" to make it work of course 😕

I'm wondering if there's a better way on doing this... 🤔

The special thing here is that we have 2 totally different stores that aren't really connected in any way (except the data is probably also in the own redux state, which it is in my case)...

Maybe @gaearon can shed his light on this one?

**Comment by dahlbyk** [comment]
Well, think of Griddle's internal Redux store as an implementation detail. You can hook into it with middleware, but it wouldn't make sense (IMO) for an external component to dispatch actions to the internal store.

From a general React standpoint, providing the current `filter` via `props` seems totally legit. Really the only "hack" here is that `filter` isn't documented as part of the [initial state](https://github.com/GriddleGriddle/Griddle/blob/250ea2a019895ee490e9f70e1ca26b4dfb9e181a/src/module.d.ts#L385-L397) that can be set via props.

**Comment by MichaelDeBoey** [comment]
@dahlbyk If you think of it like that, you're right indeed that it makes sense to provide `filter` as a prop to the `Griddle` component.

The only strange thing I see then is that it isn't shown in the `Filter`-box (like you dan see in [v2 of the CodeSandbox](https://codesandbox.io/s/k334vwn1zr?module=%2FGrid.js))... 😕

**Comment by dahlbyk** [comment]
> The only strange thing I see then is that it isn't shown in the `Filter`-box (like you dan see in v2 of the CodeSandbox)... 😕

At a glance, that seems like a bug in Griddle. `value` isn't set on the [`Filter` `input`](https://github.com/GriddleGriddle/Griddle/blob/250ea2a019895ee490e9f70e1ca26b4dfb9e181a/src/components/Filter.js#L17-L24).

**Comment by MichaelDeBoey** [comment]
@dahlbyk Glad to have caught a bug here 🙂

OK setting the `Filter` value via a prop on the `Griddle` components seems logical indeed, but retrieving it is still not possible, except if I manage that piece of state to myself...?

**Comment by MichaelDeBoey** [comment]
@dahlbyk I'm not seeing it like "a child providing state to it's parent", I'm seeing it more like an extra 'global' state, that I can access, like I do with my own Redux state.

**Comment by dahlbyk** [comment]
> I'm seeing it more like an extra 'global' state, that I can access, like I do with my own Redux state.

Except it's not global. It's mostly hidden inside the `<Griddle />`. 😀 And if you have a page with two `<Griddle />` instances, even if they use the same `storeKey`, you'll have two distinct internal stores.

If you want 'global' state, the right thing to do is to put it in your own app's Redux store.

**Comment by MichaelDeBoey** [comment]
@dahlbyk Yeah I know it's not global, but I think you understand what I'm trying to say.

Just feels a bit grim at first, but after some thoughts about it, it (like you said) an implementation detail, so as long as I'm not creating custom components that replace the default ones, I indeed should keep that piece of state in my own store and just pass it as a prop 🙂
